### ISSUE TITLE
Overflow error

### ISSUE THREAD

**Issue by abdulganij** [ReactionMechanismGenerator/RMG-Py#564]
Hi

I tried to generate a mechansim for 30 carbon molecule (n-alkane) but RMG is unable to generate a mechasnim.  Screenshot attached.

Is there an upper C no limit above which RMG will not work.
Please help.

INPUT FILE

# Data sources
database(
    thermoLibraries = ['primaryThermoLibrary'],
    reactionLibraries = [],
    seedMechanisms = [],
    kineticsDepositories = ['training'],
    kineticsFamilies = 'default',
    kineticsEstimator = 'rate rules',
)

# List of species
species(
    label='triacontane',
    reactive=True,
    structure=SMILES("CCCCCCCCCCCCCCCCCCCCCCCCCCCCCC"),
)

# Reaction systems
simpleReactor(
    temperature=(1350,'K'),
    pressure=(1.0,'bar'),
    initialMoleFractions={
        "triacontane": 1.0,
    },
    terminationConversion={
        'triacontane': 0.9,
    },
    terminationTime=(1e7,'s'),
)

simulator(
    atol=1e-16,
    rtol=1e-8,
)

model(
    toleranceKeepInEdge=0.0,
    toleranceMoveToCore=0.1,
    toleranceInterruptSimulation=0.1,
    maximumEdgeSpecies=100000,
)

options(
    units='si',
    saveRestartPeriod=None,
    generateOutputHTML=True,
    generatePlots=True,
    saveEdgeSpecies=True,
    saveSimulationProfiles=True,
)
![capture](https://cloud.githubusercontent.com/assets/17053949/12797218/dfb41bd4-cad3-11e5-9d60-b0d2793c03f4.PNG)

**Comment by connie** [comment]
So it seems that we have a unit test here: https://github.com/ReactionMechanismGenerator/RMG-Py/blob/master/rmgpy/molecule/moleculeTest.py#L1300

On my two 64 bit computers (mac and linux), I do not get any int overflow with 30 carbons, nor even with 60 carbon n-alkane.  but the error begins once I try n-C80. @nick do you see any difference in behavior with your 32-bit machine?

**Comment by rwest** [comment]
Hi.
There will always be some finite upper limit.
It seems you have found the limit in the current implementation. 
We may be able to the particular crash you found, and then the limit will be higher. Eventually the limit will just be that it takes infinitely long time to find all the many many reactions feasible with such a large molecule (but at least it shouldn't crash).

In the mean time I would recommend altering your species constraints.
http://reactionmechanismgenerator.github.io/RMG-Py/users/rmg/input.html#species-constraints.
I would set them all just a bit larger than your starting molecule, thus allowing it to fall apart.
Eg. if you're starting with C32, `maximumCarbonAtoms=35` and `maximumHeavyAtoms=40`. Otherwise it may not react.

@nickvandewiele, any thoughts to avoid the connectivity value problem?
Looks like the limits for straight chain alkanes are C35 for a 32bit integer and C79 for a 64bit long. (Branched ones may be smaller?)

```python
In [43]: for i in range(1,120):
    m = Molecule(SMILES='C'*i)
    print i, log(max([a.connectivity for a in m.atoms])) / log(2)
   ....:     
1 2.0
2 5.32192809489
3 7.02236781303
4 8.74819284959
5 8.88264304936
6 10.4377520723
7 10.4898479604
8 11.9829935747
9 12.0042204663
10 13.4723100637
11 13.4810425749
12 14.9385688271
13 14.9421479204
14 16.3952496116
15 16.3967051135
16 17.8479626625
17 17.8485495911
18 19.2990492309
19 19.2992840513
20 20.7494750175
21 20.7495683109
22 22.199634906
23 22.1996717458
24 23.6496886026
25 23.649703073
26 25.0997001885
27 25.0997058461
28 26.5496951709
29 26.5496973738
30 27.9996836409
31 27.9996844956
32 29.4496695677
33 29.4496698983
34 30.8996545054
35 30.8996546329
36 32.3496390597
37 32.3496391087
38 33.7996234658
39 33.7996234846
40 35.2496078149
41 35.2496078221
42 36.699592142
43 36.6995921447
44 38.1495764607
45 38.1495764617
46 39.5995607761
47 39.5995607765
48 41.0495450904
49 41.0495450905
50 42.4995294041
51 42.4995294042
52 43.9495137177
53 43.9495137177
54 45.3994980312
55 45.3994980312
56 46.8494823447
57 46.8494823447
58 48.2994666582
59 48.2994666582
60 49.7494509717
61 49.7494509717
62 51.1994352852
63 51.1994352852
64 52.6494195986
65 52.6494195986
66 54.0994039121
67 54.0994039121
68 55.5493882256
69 55.5493882256
70 56.9993725391
71 56.9993725391
72 58.4493568525
73 58.4493568525
74 59.899341166
75 59.899341166
76 61.3493254795
77 61.3493254795
78 62.799309793
79 62.799309793
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-43-a6f177b7182d> in <module>()
      1 for i in range(1,120):
----> 2     m = Molecule(SMILES='C'*i)
      3     print i, log(max([a.connectivity for a in m.atoms])) / log(2)
      4 

/Users/rwest/Code/RMG-Py/rmgpy/molecule/molecule.so in rmgpy.molecule.molecule.Molecule.__init__ (build/pyrex/rmgpy/molecule/molecule.c:11946)()

/Users/rwest/Code/RMG-Py/rmgpy/molecule/molecule.so in rmgpy.molecule.molecule.Molecule.fromSMILES (build/pyrex/rmgpy/molecule/molecule.c:21865)()

/Users/rwest/Code/RMG-Py/rmgpy/molecule/parser.so in rmgpy.molecule.parser.fromSMILES (build/pyrex/rmgpy/molecule/parser.c:6815)()

/Users/rwest/Code/RMG-Py/rmgpy/molecule/parser.so in rmgpy.molecule.parser.fromSMILES (build/pyrex/rmgpy/molecule/parser.c:6697)()

/Users/rwest/Code/RMG-Py/rmgpy/molecule/parser.so in rmgpy.molecule.parser.__parse (build/pyrex/rmgpy/molecule/parser.c:3275)()

/Users/rwest/Code/RMG-Py/rmgpy/molecule/parser.so in rmgpy.molecule.parser.__fromSMILES (build/pyrex/rmgpy/molecule/parser.c:2835)()

/Users/rwest/Code/RMG-Py/rmgpy/molecule/parser.so in rmgpy.molecule.parser.parse_openbabel (build/pyrex/rmgpy/molecule/parser.c:3768)()

/Users/rwest/Code/RMG-Py/rmgpy/molecule/parser.so in rmgpy.molecule.parser.fromOBMol (build/pyrex/rmgpy/molecule/parser.c:8602)()

/Users/rwest/Code/RMG-Py/rmgpy/molecule/graph.pyx in rmgpy.molecule.graph.Graph.updateConnectivityValues (build/pyrex/rmgpy/molecule/graph.c:5987)()
    291         connectivityValues = self.__update([len(vertex.edges) for vertex in self.vertices], 0)
    292         for vertex, value in zip(self.vertices, connectivityValues):
--> 293             vertex.connectivity = value
    294 
    295 

OverflowError: Python int too large to convert to C long
```

More thinking aloud: 
if the default constraints are C10, does that mean we proposed a molecule >> C10 (eg. C60) and tried to calculate its connectivityValue before imposing the species constraints? This may be another possible optimization: don't even apply reaction templates that will generate molecules you know will end up violating a constraint.

**Comment by connie** [comment]
Are you referring to the fact that we check forbidden structures after applying the reaction templates?  

I suppose we can NOT apply them, but that would require knowing what the product templates form prior to applying them. And can potentially only work for things like R_recombination where we think that 2 radicals will produce a larger combined molecule.  It will be difficult and potentially unsafe to make that prediction with other types of forbidden species constraints or other families.

I see some temporary solutions:
- Catch the overflow error and tell user that they should add species constraints or cannot run models with such large molecules
-Alter the `updateConnectivity()` algorithm to accommodate overflow errors.

**Comment by rwest** [comment]
Yes, that's what I referred to. I agree the only families when you know in advance it'll produce a molecule that violates a constraint are recombination ( A+B->C) type families. So may not be worth the pre-screening in terms of optimization.

Other possible approaches:
3. catch the overflow error, and at that point check if the species would have violated constraints anyway, in which case discard.
4. evaluate and enforce species constraints before calculating connectivity values.
5. make connectivity values that are smaller than the current ones. I think the only requirement is that they're unique and canonical, @nickvandewiele? perhaps monotonic with connectivity. Do they really need to be so large?

**Comment by nickvandewiele** [comment]
The connectivity values have never been canonical for every possible graph, not even after my algorithm changes that lead to much larger atom connectivity values. 

Currently, the atom connectivity values are only used for speed-up of graph isomorphisms (as was done in the past), while the augmented InchI functionality uses  rdkit's canonicalization algorithms.

In the past, I refrained from reverting the new connectivity value implementation, because I thought it wouldn't cause any major harm. But with the current situation I am inclined to advocate for reverting the algorithm to the previous state, in which connectivity values will never grow superlarge, to avoid situations like this.

**Comment by abdulganij** [comment]
Hi,
Thanks alot for the reply. I appreciate it.

1. I set maximumCarbonAtoms=35 and maximumHeavyAtoms=40 as suggested for the C-30 molecule. I still get the same overflow error. I use a win7 32 bit machine. If I were to switch to MAC will the problem be resolved.

2. May I know whats the difference between Model Core and edge. I read the theory guide, but I would appreciate if you could explain in simple words.

Thanks alot.

**Comment by nickvandewiele** [comment]
Hi,

The _core_ refers to the molecules and reactions that are part of the reaction mechanism because they were deemed important enough by RMG to be included.

The _edge_ refers to the molecules and reactions that are potential candidates to be added the reaction mechanism.  Whether they will be added to the reaction mechanism (the core) or not, will be determined by RMG.

The core and edge increase in size through the course of an RMG simulation. As RMG iteratively adds one species at a time in the core, many more potential candidate molecules and reactions are generated as well, and added to the edge.

Finally, most of edge molecules/reactions will not be part of the final reaction mechanism as they were not important enough to be included.

Hope this helps.

**Comment by nickvandewiele** [comment]
Hi @abdulganij :
we have worked on a solution that would (hopefully) enable you to solve your aforementioned problem. 

If you have installed RMG-Py from source, you can do the following to download and install in the latest modifications of RMG-Py:
```
cd /path/to/RMG-Py
git checkout master
git fetch origin
git reset --hard origin/master
make clean;make
```

Let us know if this works!

**Comment by abdulganij** [comment]
Hi
I installed the Basic users version, not the source code.

Anyhow I ran the program in Mac OS. It was continuously running for 7 days and 1 hour untill I stopped the program. There were 170,000 edge reactions.

Will the program converge sooner if I increase the sensitivity values. Please advise on this.

Also I selected 10 hours as the maximum time for the program to run, but it did not. It kept running until I ended it.

Thanks

**Comment by nickvandewiele** [comment]
Hi @abdulganij 

First of all: the case you are trying to simulate is very challenging for RMG. In theory, the input you defined will lead to a very large model. In practice, you will probably bump into the limits of your computer, i.e. RMG-Py will run out of memory to use, and will crash. Be aware that intermediate mechanisms are written to the `chemkin/` folder, so that you can still use the mechanisms, even when RMG-Py crashed at some point.

Some tips :
**Reduce the termination criteria**
Only keep the termination conversion, and set it to 10% rather than 90%. Keep in mind that, under this modification, only the chemistry will be included that leads to lower conversions; conversion chemistry of primary products may not necessarily be included.
```
terminationConversion={
'triacontane': 0.1,
},
```

**Increase the toleranceMoveToCore**
`toleranceMoveToCore=0.5`
This sets the bar higher to add new molecules to the mechanism.

**Prune more agressively**
For the given `toleranceMoveToCore` you can set the pruning parameters to:
```
model(
toleranceKeepInEdge=0.01,
toleranceMoveToCore=0.1,
toleranceInterruptSimulation=100000000,
maximumEdgeSpecies=100000,
)
```
Pruning proactively frees up the memory so that RMG-Py can continue longer without running into memory issues. However, it will not speed up simulations, though.

**Comment by connie** [comment]
This issue has been resolved via f9d9960ec4bd68.  See the pull request https://github.com/ReactionMechanismGenerator/RMG-Py/pull/570 for more details.
